### PR TITLE
Interactions

### PR DESCRIPTION
**Pull request by L0neGamer**
Add interactions handling, including making application commands (currently basic but can be expanded fairly easily) and working with interactions.

also updates discord-haskell to the latest version (1.12.4) and fixes the issues there.

Also also adds threads support, and made it so that tablebot automatically joins any threads that are created.

Example working here:

slashcommands

### CODE REVIEW

**Comment by finnbar** [comment]
Very quick question - I know that Sahasrara is planning to eventually be based off this and is a public bot so is present in many servers. How do interactions (ha) interact with this? I haven't read the code through yet, but given Amelie's comment I assume there's no way to turn them off? Would this PR mean that public bots are no longer allowed?

(Admittedly, I think the auth stuff is also partially incompatible with public bots since it relies on fixed role IDs. This is fine if you're happy with authorised commands only being able to be run in a given server though, or if I remember correctly Server Owner gets to bypass the auth stuff anyway. If I'm wrong, that may be another required fix for making the bot easier for others to run.)

**Comment by L0neGamer** [comment]
Currently application commands (slash commands, user commands, message commands, and modals) are only set up in a single, given server (which is required). There are two ways to solve this:

1) the server id for such application command creation is instead a list of ids, specifying which servers to add the commands to. This has the downside that each such server would need to be added to the list, which is infeasible for a public bot. ofc something can be aded which allows a server to opt in, meaning it uses the database to store opt-in servers, but that's a bit excessive because...
2) you can create global application commands.; the bot doesn't use those currently because they take up to an hour to activate, but we can combine the best of both worlds by having an absent server id meaning "create these commands globally", and then having a list present meaning "create in these servers"

Basically, there's a variety of options all of which should be fairly easy to do. 

Another change I should make is to let having these application commands be optional.

File: src/Tablebot/Plugins/Roll/Plugin.hs
Comment (finnbar) [inline src/Tablebot/Plugins/Roll/Plugin.hs:234]: Could there be a helper function that avoids all of the `Nothing` here, or is there no way to narrow it down? (Fine if not, just seems like a whole lot of Nothing.)

File: src/Tablebot/Plugins/Roll/Plugin.hs
Comment (finnbar) [inline src/Tablebot/Plugins/Roll/Plugin.hs:78]: Could maybe be a helper function `makeInteractionButton`, to avoid much use of Nothing? (Again, it might be that there are serious cases where those aren't Nothing and thus my point is not sensible.)

Also, what's with the `T.take`?

File: src/Tablebot/Plugins/Quote.hs
Comment (finnbar) [inline src/Tablebot/Plugins/Quote.hs:382]: why is this line so long

File: src/Tablebot/Plugins/Quote.hs
Comment (finnbar) [inline src/Tablebot/Plugins/Quote.hs:381]: This definition makes it look like you have to put in a hell of a lot of work to get interactions to work. Is there a way this could be reused/reduced? As I say in the conclusion of this review, I'd like to see a tutorial about interactions and slash commands - maybe when you write this tutorial it'll become clear how things could be simplified.

File: src/Tablebot/Handler.hs
Comment (finnbar) [inline src/Tablebot/Handler.hs:127]: Is this correct? I thought a blank string lead to global application commands or something?

**Review by finnbar — CHANGES_REQUESTED** [review]
The code works perfectly when testing, and provides a great user experience. I'm also astounded it still works even after some blacklisting. I have no worries about how it functions.

While I trust your code to be good, I had issues understanding it, however. When you provide a PR with _33_ files changed, you need to explain very clearly what you've implemented. I had to guess what functionality you'd actually implemented because you don't say all of the commands which benefit from slash commands / interactions, only that you can now reroll with an interaction. You also seem to have made a bunch of dice changes with `Proxy`?

As another future PRs thing - it would have been really helpful for you to explain the main types / functions being added with each feature. For example, when you talk about interactions it would have been helpful for you to point out the added types `EnvApplicationCommandRecv` and `EnvComponentRecv` - you don't need to describe them in great detail, just point them out so I know where to go first.

I want a tutorial written about Interactions before I feel comfortable merging this - a number of questions arose while I was reading the code that I think would be aided by a tutorial stepping through how to add interactions and slash commands. This is particularly pertinent because it seems like a plugin writer will have to put in significant effort if they'd like to add application commands to their plugin.

**Comment by finnbar** [comment]
Notes for when I have to get this running in the actual bot:

* Need to set `SERVER_ID` and `stack clean` in the instance running on hopper, since this requires `SERVER_ID` and does a bunch of package changes.
* Need to update Discord Intents to allow message reading.

File: src/Tablebot/Plugins/Quote.hs
Comment (L0neGamer) [reply src/Tablebot/Plugins/Quote.hs:382]: karl doesn't like short names and benji has word wrap on by default (I'll fix this)

File: src/Tablebot/Handler.hs
Comment (L0neGamer) [reply src/Tablebot/Handler.hs:127]: if the environment variable is not present, application commands will not be generated. if the variable is present but is `"global"`, global commands will be created. else, the specific named server will be used for the application commands.

File: src/Tablebot/Utility/SmartParser.hs
Comment (L0neGamer) [reply src/Tablebot/Utility/SmartParser.hs:89]: the status is "I think it all works but this stuff put me in a fugue state when I made it so I would love for someone to check properly"

File: src/Tablebot/Plugins/Quote.hs
Comment (L0neGamer) [reply src/Tablebot/Plugins/Quote.hs:381]: so the issue is that our current subcommanding turns things into the normal text-parsing straight away, which means I've had to manually implement subcommands here.  Additionally, due to how interactions work, I've had to make the quote adding and quote editing "self referential", meaning that the "view quote" link is the bot's message itself, which is obviously a bit tricky.

I also have some rudimentary slash command auto complete implemented for `/quote show`, which isn't very clean either.

**Comment by L0neGamer** [comment]
finnbar you were right
I've started the tutorial and I've seen another way of doing this
I don't think I want to implement it right now cause it's even more internals work that I don't have the time to do right now, but the gist is if we create some mkCommand functions we could have the text parsers and the application command parser stuff processed at the same time, and put in the same data structure
which would hopefully reduce on some of the separation
and if we're really smart, we could do the subcommand stuff at the same time

File: tutorials/3.Interactions.md
Comment (finnbar) [inline tutorials/3.Interactions.md:125]: It would be nice to have a tiny mention of what `unROI` is and why it is necessary

File: tutorials/3.Interactions.md
Comment (finnbar) [inline tutorials/3.Interactions.md:3]: Maybe add a screenshot here showing what these interactions are, and what the user can expect to get out of this tutorial?

File: tutorials/3.Interactions.md
Comment (finnbar) [inline tutorials/3.Interactions.md:158]: You've neglected to say what labelling _is_ at this point - another screenshot showing a slash command at work may help?

File: tutorials/3.Interactions.md
Comment (finnbar) [inline tutorials/3.Interactions.md:166]: What is `Labelled`? Is it provided by Tablebot or discord-haskell? I'm guessing it's a constructor for a labelled argument, but it'd still be nice to see that in a sentence - it groups a value with some label.

File: tutorials/3.Interactions.md
Comment (finnbar) [inline tutorials/3.Interactions.md:203]: ~~man and I thought the _first_ tutorial was contrived lol~~

File: tutorials/3.Interactions.md
Comment (finnbar) [inline tutorials/3.Interactions.md:230]: Please provide an example of this - I'm not exactly sure what you're referring to here.

**Review by finnbar — CHANGES_REQUESTED** [review]
I'm happy with the code and how it runs, I just have a few points for your tutorial. This can be merged once these are fixed, merge conflicts are resolved, and I quickly test the results of that.

**Pull request merged**
